x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/job/config/AnalysisConfig.java

```diff
@@ -336,6 +338,13 @@ public List<String> fields() {
         return collectNonNullAndNonEmptyDetectorFields(Detector::getFieldName);
     }
 
+    public String maybeTruncateCatgeorizationField(String categorizationField) {
+        if (termFields().contains(categorizationFieldName) == false) {
```

Comment:
This is really inefficient because `termFields()` is building a `Set` from scratch each time it's called, and then that `Set` becomes garbage after just one lookup.

It needs to be changed to build the set once and cache it.  That probably means moving this logic into the `AbstractDataToProcessWriter` class, which could cache the term fields on construction.